src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -95,16 +101,21 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
     def checkCq(cq: ConjunctiveQuery) {
       checkAtom(cq.head)
-      cq.bodies foreach { x => x foreach checkAtom }
+      cq.bodies foreach { x =>
```

Comment:
You could define a higher order iterator for bodies that applies the checkers plugged into it to `Atom`s, `ModifiedAtom`s, etc. and reuse it from the places where you do `foreach` over bodies.  Instead of defining multiple iterators, you could have a single iterator that accepts optional checkers for each Body type that defaults to no-op, then visits all nodes in the parse tree.